### ISSUE TITLE
Duplicate child components break rendering of the parent component

### ISSUE THREAD

**Issue by kevinmpowell** [EightShapes/Blocks#108]
`<div data-component="parent" data-variation="default"></div>`

**parent.html**
````
<section data-variation="default">
    <div class="parent-class">
        <div data-component="child" data-variation="default"></div>
        <div data-component="child" data-variation="default"></div>
    </div>
</section>
````

**child.html**
````
 <section data-variation="default">
    <h1>Hello World!</h1>
</section>
````

## Expected Output
````
    <div class="parent-class">
        <h1>Hello World!</h1>
        <h1>Hello World!</h1>
    </div>
````

## Actual Output
The component doesn't even render. It fails somewhere around line 359, https://github.com/EightShapes/Blocks/blob/master/js/jquery.8SblocksLoader.js#L359

Somehow the child rendered count isn't getting incremented correctly and thus the parent component doesn't proceed with rendering.

@humancompanion, I dug and dug on this trying to find the issue and I kept getting lost in the recursion. Do you have any ideas off the top of your head?

**Comment by humancompanion** [comment]
I was rendering the parent template before setting the UUID on the children. Having fixed that this is now working. 

Also, finally gave up and worked around Jasmine's async behavior with setTimeout(). Triggers were not working. Thus there is now a test of this behavior.

**Comment by humancompanion** [comment]
Now that I'm on a project that sets components.replace_reference = true globally I've found that this change broke rendering in that code block in a number of nested instances. I'm also realizing we don't really have a good testing setup for config settings like this one.

**Comment by humancompanion** [comment]
This problem now comes from this line in childDoneRendering(child) in blocks-component:

````
self.$el.find('[data-component]');
````

When the parent element is being replaced (either data-place="replace" or components.replace_reference = true in the config) that find will fail. It fails because self.$el has already been replaced in renderTemplate() in the self.replace_reference === true block.

I've fixed that issue by doing the find, checking the results, and handling the null/undefined case. However, testing revealed another bug. I'm going to fully document this new bug because I've hit and solved this before and I'm a little ticked that this problem keeps seeping back in somehow. The problem is that in renderTemplate() when using replace we do this:

````
self.$el = $(rendered_tmpl);
````

rendered_tmpl is self.template(self.template_data) which is 
````
handlebars.compile(tmpl);
````

For some reason the text string that handlebars returns gets converted into an array by jquery when we wrap it in $(). 

I thought I fixed this problem with this line before handing it to handlebars:

````
 tmpl = $.trim(variation_html).replace(/\n\s*/g, '');
````

But, alas, jquery and handlebars still find a way to mess it up. So now I'm onto how to detect the problem and concat this array back together again.
